### PR TITLE
Headless electron

### PR DESCRIPTION
**Pull request by Autarc**
Using [electron-webrtc](https://github.com/mappum/electron-webrtc) as an alternative engine for [wrtc](https://github.com/js-platform/node-webrtc) (see #5).

### CODE REVIEW

File: lib/global.js [hunk 1]
```diff
@@ -1,14 +1,3 @@
-var createTorrent = require('create-torrent')
-var wrtc = require('wrtc')
+var wrtc = require('electron-webrtc')()
 
 global.WRTC = wrtc
-
-global.WEBTORRENT_ANNOUNCE = createTorrent.announceList
```
Comment (rom1504) [inline lib/global.js]: why did you remove this ? It makes it possible to announce the (public) torrent to websocket trackers too (so browsers can download the torrents from webtorrent-hybrid peers)

Comment (Autarc) [inline lib/global.js]: The current version overwrote the [default trackers of createTorrent](https://github.com/feross/create-torrent/blob/master/index.js#L4), so only the ones which support the websocket protocol kept enabled. Using a chromium based implementation instead of `wrtc` it should expect the same support as the regular browser version.

Comment (rom1504) [inline lib/global.js]: No. This is there to add the websockets default trackers of createTorrent to any torrents. 
Without it, it will only use the trackers of the (existing) torrents, which mean for most torrents it won't announce to any websocket trackers, and won't be available from browsers.

(see https://github.com/feross/webtorrent-app/blob/c176e3897ac70d55acd099a4f15464fb85e79ed7/renderer/index.js#L34 )

**Comment by rom1504** [comment]
I tested this (./bin/cmd seed README.md and putting the infohash in a browser) and it doesn't seem to work. I'm not getting any error though.
Does it work for you ?

File: lib/global.js (hunk 1 above)
Comment (Autarc) [inline lib/global.js]: Thought the default trackers will be added to the torrents anyway (see `announceList`). Is there a reason why websocket trackers should be forced only at standalone versions like the app/hybrid and not the browser ?

Comment (rom1504) [inline lib/global.js]: they are forced in the browser too https://github.com/feross/instant.io/blob/c8924fe4cea665c938911edf5b7efaa2c4b1e9ab/client/index.js#L15

Only place where they aren't forced is webtorrent-cli, because it doesn't support webrtc anyway.

> Thought the default trackers will be added to the torrents anyway (see announceList).

They will be added to new torrents you create with createTorrent without giving a list of trackers manually. But they won't be added to existing torrents you download (/seed), which is the reason why this global.WEBTORRENT_ANNOUNCE exists

Comment (Autarc) [inline lib/global.js]: Okay, its a bit confusing that all of the webtorrent projects use the [create-torrent](https://github.com/feross/create-torrent) module and each of them still handles this restriction by themself. What do you recommend, reverting the removal here or adding a reference at the shared source ?

Comment (rom1504) [inline lib/global.js]: I guess it might make sense to check if there is webrtc support and in that case to add these torrents, in webtorrent itself.
That would make it possible to remove all these adding of trackers in these projects.
> still handles this restriction by themself

Really, it's not a restriction. If you don't define WEBTORRENT_ANNOUNCE , it doesn't add any trackers at all.

So yeah I think either revert the removal here, or add it in webtorrent itself (or both until it's merged and published in webtorrent).

Comment (Autarc) [inline lib/global.js]: > Really, it's not a restriction. If you don't define WEBTORRENT_ANNOUNCE , it doesn't add any trackers at all.

Sorry, with "restriction" I always meant filtering the list of already defined trackers - not the actual logic. Be aware of the reason using `WEBTORRENT_ANNOUNCE`...

Comment (rom1504) [inline lib/global.js]: Yes it filters the list because there's no need to add normal trackers to existing torrents, because they usually have them already.

**Comment by Autarc** [comment]
Yeah the `webtorrent-cli` commands work with the hybrid package for both environments (browser & node). I will update the built as soon as these changes are available through its dependencies:
- [bittorrent-tracker](https://github.com/feross/bittorrent-tracker/pull/126): checkout the fix to properly close the webtorrent instance without a leak
- [electron-webrtc](https://github.com/mappum/electron-webrtc/commit/605af48f791b5e1b77ed79ef0acbfa4f0929f66f): @mappum publish soon a new npm version including the latest `electron-eval`
- [webtorrent-cli](https://github.com/feross/webtorrent-cli/pull/1): checkout the increased delay to ensure rendering on the commandline

**Comment by mappum** [comment]
@Autarc Just published `electron-webrtc@0.0.4`

**Comment by mappum** [comment]
It might be helpful to copy the info about running `electron-webrtc` in headless mode into the README: https://github.com/mappum/electron-webrtc#running-on-a-headless-server

**Comment by Autarc** [comment]
@mappum It was already mentioned [here](https://github.com/Autarc/webtorrent-hybrid/commit/2b3d7b2a242aab588da0653e8650f17da4147df1). In addition it comes now with a shell script to ease automating the process :)

File: lib/global.js (hunk 1 above)
Comment (rom1504) [inline lib/global.js]: https://github.com/feross/webtorrent/issues/672#issuecomment-196614860

Comment (Autarc) [inline lib/global.js]: > WebTorrent.app and webtorrent-hybrid should probably stop adding trackers to added files as well 

So it will be kept removed for now to avoid adding 3rd party endpoints to the torrents.

File: package.json [hunk 2]
```diff
@@ -39,6 +38,7 @@
     "url": "git://github.com/feross/webtorrent-hybrid.git"
   },
   "scripts": {
+    "check": "npm outdated -depth 0",
```
Comment (feross) [inline package.json]: Please remove this.

File: package.json [hunk 3]
```diff
@@ -12,13 +12,12 @@
     "url": "https://github.com/feross/webtorrent-hybrid/issues"
   },
   "dependencies": {
-    "create-torrent": "^3.22.0",
-    "webtorrent": "0.x",
-    "webtorrent-cli": "1.x",
-    "wrtc": "0.0.x"
+    "electron-webrtc": "^0.1.0",
+    "webtorrent": "^0.85.1",
```
Comment (feross) [inline package.json]: Please keep the webtorrent version set to `0.x` so we pull in minor version updates.

For example, `^0.85.1` will not install `0.86.0` or `0.87.0`. Semver is confusing for versions less than 1.0.0. We'll fix this by releasing webtorrent 1.0.0 very soon.

File: package.json [hunk 4]
```diff
@@ -12,13 +12,12 @@
...
+    "webtorrent-cli": "^1.0.1"
```
Comment (feross) [inline package.json]: Changing the others to use `^` is fine.

File: README.md [hunk 5]
```diff
@@ -5,11 +5,13 @@
 [![NPM Downloads][webtorrent-downloads-image]][webtorrent-downloads-url]
 
 
-### Streaming torrent client for node & the browser
+### Streaming torrent client for node environments
```
Comment (feross) [inline README.md]: Thanks for fixing this.

File: README.md [hunk 6]
```diff
@@ -5,11 +5,13 @@
...
+While previous versions were using [wrtc](https://github.com/js-platform/node-webrtc), a WebRTC implementation via native extensions, for better compatibility the current one is based on [electron-webrtc](https://github.com/mappum/electron-webrtc). Its creating a hidden electron process which communicates with the Chromium API in the background. Since it comes with the overhead of [electron-prebuilt](https://github.com/electron-userland/electron-prebuilt) we are looking further  for a seamless integration through other implementations like [Node-RTCPeerConnection](https://github.com/nickdesaulniers/node-rtc-peer-connection).
+
+To run this package on a headless server execute [the provided script](bin/prepareHeadless.sh) or follow [these instructions](https://github.com/mappum/electron-webrtc#running-on-a-headless-server).
```
Comment (feross) [inline README.md]: Thanks for updating the readme.

File: bin/prepareHeadless.sh [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  fi
+
+  running=$(pidof Xvfb)
+  if [ -z "$running" ]; then
+    kill running
+  fi
+
+  export DISPLAY="0:99"
+  Xvfb :99 -screen 0 1024x768x24 > /dev/null 2>&1 &
+fi
```
Comment (feross) [inline bin/prepareHeadless.sh]: This is very handy

File: lib/global.js [hunk 8]
```diff
@@ -1,14 +1,18 @@
...
 
 global.WRTC = wrtc
 
-global.WEBTORRENT_ANNOUNCE = createTorrent.announceList
-  .map(function (arr) {
-    return arr[0]
-  })
-  .filter(function (url) {
-    return url.indexOf('wss://') === 0 || url.indexOf('ws://') === 0
-  })
```
Comment (feross) [inline lib/global.js]: Please re-add this. See https://github.com/feross/webtorrent/issues/672#issuecomment-196614860 for reasoning.

Eventually, we'll remove this from `webtorrent-hybrid` and [`webtorrent-app`](https://github.com/feross/webtorrent-app), but we need to keep it for now.

File: lib/global.js (hunk 1 above)
Comment (feross) [inline lib/global.js]: Sorry if I wasn't clear enough in my comment. I'd like to keep automatically adding wss:// trackers to all torrents in `webtorrent-hybrid` for the time being. Eventually, once folks start adding wss:// trackers to .torrent files, we can remove it.

If we remove it now, then torrents downloaded with `webtorrent-hybrid` will not be seeded to web peers by default.

**Comment by feross** [comment]
@Autarc Merged your PR: https://github.com/feross/webtorrent-cli/pull/1

**Comment by feross** [comment]
@Autarc It's not necessary to publish new versions of `torrent-discovery` and `webtorrent` to get the latest changes. That's only necessary when we publish new major versions (X.0.0). That's the nice thing about the caret `^` – you get minor (0.X.0) and patch (0.0.x) versions for free the next time you `rm -rf node_modules && npm install`

**Comment by feross** [comment]
Wow, just tried this out. It works quite well!

I'm going to merge it now, and make the changes I suggested inline right now.

**Comment by feross** [comment]
Released as 1.0.0.

**Comment by feross** [comment]
@Autarc I added you as a collaborator on this repo. Contributing guidelines are here: https://github.com/feross/webtorrent/blob/master/CONTRIBUTING.md Cheers!
